llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -3047,24 +3049,31 @@ static void getUnsignedMonotonicValues(SmallPtrSetImpl<Value *> &Res, Value *V,
   if (Type == MonotonicType::GreaterEq) {
     if (match(I, m_Or(m_Value(X), m_Value(Y))) ||
         match(I, m_Intrinsic<Intrinsic::uadd_sat>(m_Value(X), m_Value(Y)))) {
-      getUnsignedMonotonicValues(Res, X, Type, Depth);
-      getUnsignedMonotonicValues(Res, Y, Type, Depth);
+      getUnsignedMonotonicValues(Res, X, Type, Q, Depth);
+      getUnsignedMonotonicValues(Res, Y, Type, Q, Depth);
+    }
+    // X * Y >= X --> true
+    if (match(I, m_NUWMul(m_Value(X), m_Value(Y)))) {
+      if (isKnownNonZero(X, Q))
```

Comment:
Do we need the non-constant support?

Reply:
It is needed by assimp:
https://github.com/assimp/assimp/blob/2bce77f909b9f56c8c67cff33cc73a244505aa46/code/AssetLib/glTF2/glTF2Asset.inl#L897-L903
```
  %switch.gep = getelementptr inbounds nuw [7 x i64], ptr @switch.table._ZN5glTF28Accessor6Sparse9PatchDataEj, i64 0, i64 %307
  %switch.load = load i64, ptr %switch.gep, align 8
  %308 = mul nuw nsw i64 %switch.load, %202
  store i64 %308, ptr %20, align 8
  %309 = icmp samesign ult i64 %308, %202
```